### PR TITLE
Fix "/host unmount failure" during reboot 

### PR DESCRIPTION
**Pull request by rkdevi27**
**- Why I did it**
During /sbin/reboot, the /host unmount failure happened.
**- How I did it**
The /var/log is mounted as a separate ext4 filesystem. This ext4 filesystem is associated with a loop device and mounted along with root partition.This loop device association needs to be detached after unmounting /var/log which systemd is not performing. 
The script to stop the journal services, unmount /var/log and detele the loop device has been invoked at the closure of syslog.socket.

**- How to verify it**
I have verified the 100 iterations of reboot and confirmed issue has been fixed in all dell platforms.
Once the device is up, have checked if the /var/log, host partitions has been mounted properly.
Fast-boot, warm-boot and normal reboot has also been verified.

**- Description for the changelog**

Fixed the /host unmount failure issue.
Attaching the logs of 100 iterations in S6100 device.

**- A picture of a cute animal (not mandatory but encouraged)**

[logs_S6100.txt](https://github.com/Azure/sonic-buildimage/files/4597483/logs_S6100.txt)

### CODE REVIEW

**Comment by lguohan** [comment]
retest broadcom please

**Comment by lguohan** [comment]
retest vsimage please

File: files/image_config/syslog/host_umount.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    systemctl stop systemd-journald.socket
+    systemctl stop systemd-journald-audit.socket
+    systemctl stop systemd-journald-dev-log.socket
+}
+
+delete_loop_device() {
+    umount /var/log
+    if [[ $? -ne 0 ]]
+    then
+	exit 0
```
Comment (qiluo-msft) [inline files/image_config/syslog/host_umount.sh:17]: >	 [](start = 0, length = 1)

Do not mix space and tab. We prefer space. Could you also reformat the whole file?

File: files/image_config/syslog/host_umount.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    fi
+    losetup -d /dev/loop1
+}
+
+case "$1" in
+    journal_stop|delete_loop_device)
+	$1
+	;;
+    *)
+	echo "Usage: $0 {journal_stop|delete_loop_device}"
```
Comment (qiluo-msft) [inline files/image_config/syslog/host_umount.sh:27]: >echo [](start = 1, length = 4)

echo to stderr, and exit with a non-zero code

File: files/build_templates/sonic_debian_extension.j2 [hunk 3]
```diff
@@ -216,6 +216,11 @@ sudo cp $IMAGE_CONFIGS/rsyslog/rsyslog.conf.j2 $FILESYSTEM_ROOT_USR_SHARE_SONIC_
 sudo cp $IMAGE_CONFIGS/rsyslog/rsyslog.d/* $FILESYSTEM_ROOT/etc/rsyslog.d/
 echo "rsyslog-config.service" | sudo tee -a $GENERATED_SERVICE_FILE
 
+#copy syslog override files
```
Comment (qiluo-msft) [inline files/build_templates/sonic_debian_extension.j2:219]: >#copy [](start = 0, length = 5)

Better to follow the code style in the same file. We may need more space between # and character to have better breath.

File: files/image_config/syslog/host_umount.sh (hunk 1 above)
Comment (rkdevi27) [reply files/image_config/syslog/host_umount.sh:17]: Accepted

File: files/build_templates/sonic_debian_extension.j2 (hunk 3 above)
Comment (rkdevi27) [reply files/build_templates/sonic_debian_extension.j2:219]: Accepted

File: files/image_config/syslog/host_umount.sh (hunk 2 above)
Comment (rkdevi27) [reply files/image_config/syslog/host_umount.sh:27]: Hi Qi,

This is for line number 17 or 27?

Comment (qiluo-msft) [reply files/image_config/syslog/host_umount.sh:27]: For this line
```
echo "Usage: $0 {journal_stop|delete_loop_device}"
```

---
In reply to: [423504781](https://github.com/Azure/sonic-buildimage/pull/4558#discussion_r423504781) [](ancestors = 423504781)

Comment (rkdevi27) [reply files/image_config/syslog/host_umount.sh:27]: Accepted

**Comment by qiluo-msft** [comment]
Do you have new commit for reviewing?

**Comment by rkdevi27** [comment]
Hi Qi,

I'm working on the changes. I will commit it soon.

**Comment by joyas-joseph** [comment]
Here's an observation.

In stretch based builds:
```
root@sonic:~# which reboot
/usr/bin/reboot
root@sonic:~# 
root@sonic:~# reboot
requested COLD shutdown
/var/log: 3.9 GiB (4134080512 bytes) trimmed
/boot: 8.7 GiB (9333067776 bytes) trimmed
Thu May 14 03:07:19 UTC 2020 Rebooting with platform x86_64-dell_s6000_s1220-r0 specific tool ...
```

In buster based builds:
```
root@sonic:~# which reboot
/usr/sbin/reboot
root@sonic:~# file /usr/sbin/reboot
/usr/sbin/reboot: symbolic link to /bin/systemctl
root@sonic:~# 
root@sonic:~# 
root@sonic:~# reboot
[ 1805.643776] kdump-tools[11990]: Stopping kdump-tools: unloaded kdump kernel.
[ 1809.078008] i2c i2c-1: delete_device: Deleting device 24c02 at 0x50
[ 1809.090970] i2c i2c-1: delete_device: Deleting device 24c02 at 0x51
[ 1809.098838] i2c i2c-1: delete_device: Deleting device dni_dps460 at 0x58
[ 1809.106068] i2c i2c-1: delete_device: Deleting device dni_dps460 at 0x59
[  OK  ] Unmounted /var/lib/docker/…3e104c5c2af56de201/mounts/shm.
```

We do really want the reboot to happen through the platform specific tool.

**Comment by rkdevi27** [comment]
Hi Joyas,

In S6100 and Z9100, platform reboot is happening using systemd services to ensure graceful unmount.
Please refer PR [https://github.com/Azure/sonic-buildimage/pull/2912]
In that devices, normal reboot command will also have /host unmount issue.

**Comment by rkdevi27** [comment]
> Do you have new commit for reviewing?

Hi Qi,

committed the last changes addressing the review comments.

File: files/image_config/syslog/override.conf [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[Unit]
+After=var-log.mount host.mount
+
+[Socket]
+ExecStopPre=/usr/bin/host_umount.sh journal_stop
+ExecStopPost=/usr/bin/host_umount.sh delete_loop_device
+
```
Comment (qiluo-msft) [inline files/image_config/syslog/override.conf:7]: remove extra empty line

Comment (rkdevi27) [reply files/image_config/syslog/override.conf]: Done and committed.

**Comment by rkdevi27** [comment]
Qi/Joyas,

Please let us know if there is there any further queries on this for the merge to happen.

**Comment by qiluo-msft** [comment]
Why the title contains 'DellEMC', I think it is a general fix.

**Comment by rkdevi27** [comment]
Yeah Right. I will change it..

**Comment by joyas-joseph** [comment]
> Qi/Joyas,
> 
> Please let us know if there is there any further queries on this for the merge to happen.

Looks good to me.

**Pull request merged**